### ISSUE TITLE
Suspicious near-duplicate reactions

### ISSUE THREAD

**Issue by Devlin-Moyer** [SysBioChalmers/Human-GEM#561]
I noticed a number of pairs of reactions that had identical or near-identical stoichiometry and/or GPRs where one was a subset of the other, and wrote some Python code to identify all such pairs and use some heuristics to decide which member of each pair should be kept and which should be dropped

<details>
  <summary>Python code (it's kind of long)</summary>

```python
import cobra
import itertools as it
import numpy as np
from fractions import gcd
from functools import reduce

model = cobra.io.read_sbml_model('Human-GEM.xml')
# loop over all pairs of reactions
bad_rxns = list()
better_rxns = list()
for (r1, r2) in it.combinations(model.reactions, 2):
    # check if they involve exactly the same metabolites
    if set(r1.metabolites.keys()) == set(r2.metabolites.keys()):
        # to check if all the stoichiometric coefficients in one reaction
        # are n * the corresponding stoichiometric coefficient in the other
        # for any n, make a matrix with one column for each reaction and one
        # row for each metabolite and see what the rank of that matrix is
        stoich_mat = np.array([
            (r1.metabolites[met], r2.metabolites[met])
            for met in r1.metabolites.keys()
        ])
        if np.linalg.matrix_rank(stoich_mat) == 1:
            # see if all stoichiometric coefficients are the same magnitude
            # but opposite sign
            sign_diff_only = all(
                r1.metabolites[met] == -r2.metabolites[met]
                for met in r1.metabolites.keys()
            )
            if (not r1.reversibility) and (not r2.reversibility):
                # if they're both irreversible, see if the only difference
                # is that they're oriented in opposite directions
                sign_diff_only = all(
                    r1.metabolites[met] == -r2.metabolites[met]
                    for met in r1.metabolites.keys()
                )
                # only need to do something if there are more differences
                if not sign_diff_only:
                    # turns out all of these cases have identical GPRs but
                    # but one set of stoichiometric coefficients with a
                    # greatest common divisor that's larger than 1
                    r1_stoichs = [
                        int(abs(x)) for x in r1.metabolites.values()
                    ]
                    r1_gcd = reduce(gcd, r1_stoichs)
                    r2_stoichs = [
                        int(abs(x)) for x in r2.metabolites.values()
                    ]
                    r2_gcd = reduce(gcd, r2_stoichs)
                    # drop the reaction with the GCD that isn't 1
                    if r1_gcd > 1:
                        bad_rxns.append(r1)
                        better_rxns.append(r2)
                    elif r2_gcd > 1:
                        bad_rxns.append(r2)
                        better_rxns.append(r1)
            elif r1.reversibility and r2.reversibility:
                # if they're both reversible, see if they have the same GPR
                if r1.gene_reaction_rule == r2.gene_reaction_rule:
                    # turns out all of the cases where both have the same
                    # GPR & metabolites also have the same coefficients
                    # but are oriented in opposite directions, so pick one
                    r1_num = int(r1.id.lstrip('MAR'))
                    r2_num = int(r2.id.lstrip('MAR'))
                    # drop the one with the higher ID number
                    if r1_num > r2_num:
                        bad_rxns.append(r1)
                        better_rxns.append(r2)
                    else:
                        bad_rxns.append(r2)
                        better_rxns.append(r1)
                else:
                    # if the only difference is their GPRs, drop the one
                    # with fewer genes in the GPR
                    if len(r1.genes) < len(r2.genes):
                        bad_rxns.append(r1)
                        better_rxns.append(r2)
                    else:
                        bad_rxns.append(r2)
                        better_rxns.append(r1)
            else:
                # if one is reversible and one is irreversible, go with
                # the reversible one
                if r1.reversibility:
                    bad_rxns.append(r2)
                    better_rxns.append(r1)
                elif r2.reversibility:
                    bad_rxns.append(r1)
                    better_rxns.append(r2)
  ```
</details>

### Reactions that are completely identical except for which metabolites are "products" and which are "reactants" (which is irrelevant because both are reversible):
- [ ] Either `MAR02608` or `MAR04562` should be removed:
  - They both have identical metabolites & stoichiometry: `MAM02348c + MAM02495c <=> MAM01597c + MAM02973c`
  - They both have identical GPRs: `ENSG00000110090 or ENSG00000169169 or ENSG00000205560`
  - The only difference is the orientation: the products of `MAR02608` are the reactants of `MAR04562` and vice versa

### Reactions with identical metabolites & stoichiometry, but only one has a GPR:
- [ ] `MAR00464: MAM02685c <=> MAM02685r`, no GPR, should be removed:
  - `MAR07178` has identical metabolites & stoichiometry and actually has a GPR: `ENSG00000163586`
- [ ] `MAR11310: MAM01597c + MAM02639c <=> MAM02348c + MAM02647c`, no GPR, should be removed:
  - `MAR00108` has identical metabolites & stoichiometry and actually has a GPR: `ENSG00000110090 or ENSG00000169169 or ENSG00000205560`
- [ ] `MAR10410: MAM00270c <=> MAM00270e`, no GPR, should be removed:
  - `MAR02404` has identical metabolites & stoichiometry and actually has a GPR: `ENSG00000184999`
- [ ] `MAR02490: MAM02990m + MAM03130m <=> MAM01756m + MAM02040m + MAM02666m`, no GPR, should be removed
  - `MAR06621` has identical metabolites & stoichiometry, but actually has a GPR: `(ENSG00000048392 and ENSG00000100348 and ENSG00000167325 and ENSG00000171848) or (ENSG00000048392 and ENSG00000136810 and ENSG00000167325 and ENSG00000171848)`

### Reactions with identical metabolites, but one has all stoichiometric coefficients = 2 or 3 and one has all = 1:
- [ ] `MAR08985: 3.0 MAM01961c + 3.0 MAM02040c --> 3.0 MAM01169c + 3.0 MAM02039c`, GPR: `ENSG00000130313` should be removed
  - All stoichiometric coefficients are +/-3
  - `MAR04625` has same metabolites but all stoichiometric coefficients = +/-1 and same GPR
- [ ] `MAR08992: 2.0 MAM02846c <=> 2.0 MAM01761c`, GPR: `ENSG00000197713` should be removed:
  - All stoichiometric coefficients are +/-2
  - Same metabolites as `MAR04477`, but oriented in opposite directions, and `MAR04477` has all stoichiometric coefficients = +/-1
  - GPR of `MAR04477` is `ENSG00000197713 or ENSG00000235376`, which seems more complete

### Reactions with identical reactions & stoichiometry, but one is reversible and one is irreversible:
- [ ] Either `MAR03933` or `MAR20045` should be removed:
  - They both have identical metabolites & stoichiometry: `MAM02152c + MAM02552c <=> MAM01837c + MAM02039c + MAM02553c`
  - They both have identical GPRs: `ENSG00000197894`
  - `MAR03933` is reversible and `MAR20045` is irreversible
- [ ] Either `MAR06644` or `MAR20004` should be removed, and whichever one is kept might need an updated GPR:
  - They both have identical metabolites & stoichiometry: `MAM01230c + MAM02039c + MAM02553c <=> MAM01232c + MAM02552c`
  - `MAR20004` is irreversible but `MAR06644` is reversible
  - `MAR20004`'s GPR is just `ENSG00000198099`, but MAR06644 has a much longer GPR that does not contain that gene: `ENSG00000025423 or ENSG00000135437 or ENSG00000139547 or ENSG00000140522 or ENSG00000157326 or ENSG00000170786 or ENSG00000187630 or ENSG00000265203`

**Comment by haowang-bioinfo** [comment]
this is great!

would suggest to remove following ones:
- MAR04562
- MAR00464
- MAR11310
- MAR10410
- MAR02490
- MAR08985
- MAR08992

In addition, might be good to include the code into to `sanityCheck`.

**Comment by mihai-sysbio** [comment]
This discussion would be a good opportunity to revisit the issues previously created by @johan-gson: #347, #348 and #400.

**Comment by Devlin-Moyer** [comment]
oh I'm not sure why it didn't get caught by the Python code I used to get all the others, but I just happened across another similarly suspicious pair of reactions:
MAR00025: MAM01268n -> MAM01268c
MAR00030: MAM01286c -> MAM01286n
They have _very_ different GPRs, so I don't know what's up with that:
MAR00025: `ENSG00000065833 or ENSG00000105398 or ENSG00000112972 or ENSG00000130304 or ENSG00000139278 or ENSG00000143344 or ENSG00000147872 or ENSG00000170485 or ENSG00000198431`
MAR00030: `ENSG00000004799 or ENSG00000005471 or ENSG00000072310 or ENSG00000099194 or ENSG00000100979 or ENSG00000105398 or ENSG00000110090 or ENSG00000113790 or ENSG00000117054 or ENSG00000118137 or ENSG00000130304 or ENSG00000134240 or ENSG00000134824 or ENSG00000135218 or ENSG00000138823 or ENSG00000147872 or ENSG00000149485 or ENSG00000151726 or ENSG00000156096 or ENSG00000157184 or ENSG00000158874 or ENSG00000161533 or ENSG00000163586 or ENSG00000165029 or ENSG00000166819 or ENSG00000166821 or ENSG00000167114 or ENSG00000167588 or ENSG00000167772 or ENSG00000169710 or ENSG00000175445 or ENSG00000177666 or ENSG00000185000 or ENSG00000187048 or ENSG00000278540`
Some of the genes are in both (ENSG00000130304, ENSG00000147872, ENSG00000105398, ENSG00000112972), but most are only in one

**Comment by haowang-bioinfo** [comment]
might be better to move this block to a new issue @Devlin-Moyer

**Comment by feiranl** [comment]
> this is great!
> 
> would suggest to remove following ones:
> 
> * MAR04562
> * MAR00464
> * MAR11310
> * MAR10410
> * MAR02490
> * MAR08985
> * MAR08992
> * MAR03933
> 
> But not sure about the case of MAR06644 vs. MAR20004, can you please double check this @feiranl
> 
> In addition, it might be good to include the code into to `sanityCheck`.

Checked result:
would suggest to remove following ones:
- [x] MAR04562 
- [x] MAR00464
- [x] MAR11310
- [x] MAR10410
- [x] MAR02490
- [x] MAR08985
- [x] MAR08992
- not to remove MAR03933 since MAR20045 has been removed already.
- not to remove MAR06644, since MAR20004 has been removed already.

**Comment by haowang-bioinfo** [comment]
it seems MAR20045 is still in `develop`, please remind me which PR has the removal.

**Comment by feiranl** [comment]
Sorry for the mistake. I checked MetabolicAtlas and they aren't there. After double check, I found I added those rxns in last commit, it is ok to remove them. I modified the last comment.

**Comment by feiranl** [comment]
After double check, here are other cases also should be fixed:

MAR04521   glucose-6-phosphate + H2O   ->  glucose + Pi ENSG00000131482 or ENSG00000141349 or ENSG00000152254
MAR20020    glucose-6-phosphate + H2O   ->  glucose + Pi ENSG00000131482 or ENSG00000141349

MAR06981    4-trimethylammoniobutanal + H2O + NAD+  ->  gamma-butyrobetaine + 2 H+ + NADH ENSG00000143149
MAR20014    4-trimethylammoniobutanal + H2O + NAD+  ->  gamma-butyrobetaine + 2 H+ + NADH ENSG00000143149

MAR04689    4-aminobutanal + H2O + NAD+     ->  4-aminobutyrate + 2 H+ + NADH ENSG00000143149
MAR20015    4-aminobutanal + H2O + NAD+     ->  4-aminobutyrate + 2 H+ + NADH ENSG00000143149

MAR06631    NAD+ + retinol  ->  H+ + NADH + retinal ENSG00000025423 or ENSG00000135437 or ENSG00000147576 or ENSG00000170786 or ENSG00000172955 or ENSG00000187758 or ENSG00000196344 or ENSG00000196616 or ENSG00000197894 or ENSG00000198099 or ENSG00000248144
MAR20002    NAD+ + retinol  ->  H+ + NADH + retinal ENSG00000198099 or ENSG00000196616

MAR06646    H2O + NAD+ + retinal    ->  2 H+ + NADH + retinoate ENSG00000128918 or ENSG00000165092
MAR20012    H2O + NAD+ + retinal    ->  2 H+ + NADH + retinoate ENSG00000184254

MAR04545    5-hydroxy-L-tryptophan + H+     ->  CO2 + serotonin ENSG00000132437
MAR20090    5-hydroxy-L-tryptophan + H+     ->  CO2 + serotonin ENSG00000132437

MAR03984    2 H2O2  ->  2 H2O + O2 ENSG00000121691 or ENSG00000166016
MAR20066    2 H2O2  ->  2 H2O + O2 ENSG00000121691

MAR00067    H2O + NAD+ + 2(S)-pristanal     ->  2 H+ + NADH + pristanic acid ENSG00000072210
MAR20052    H2O + NAD+ + 2(S)-pristanal     ->  2 H+ + NADH + pristanic acid ENSG00000072210

suggest:
[ ] remove MAR20020
[ ] remove MAR20014
[ ] remove MAR20015
[ ] remove MAR20002
[ ] remove MAR20012
[ ] remove MAR20090
[ ] remove MAR20066
[ ] remove MAR20052

**Comment by Devlin-Moyer** [comment]
I believe I recently identified all of those and more in #562

**Comment by haowang-bioinfo** [comment]
> After double check, here are other cases also should be fixed:
 MAR04521   glucose-6-phosphate + H2O   ->  glucose + Pi ENSG00000131482 or ENSG00000141349 or ENSG00000152254
MAR20020    glucose-6-phosphate + H2O   ->  glucose + Pi ENSG00000131482 or ENSG00000141349
MAR06981    4-trimethylammoniobutanal + H2O + NAD+  ->  gamma-butyrobetaine + 2 H+ + NADH ENSG00000143149
MAR20014    4-trimethylammoniobutanal + H2O + NAD+  ->  gamma-butyrobetaine + 2 H+ + NADH ENSG00000143149
MAR04689    4-aminobutanal + H2O + NAD+     ->  4-aminobutyrate + 2 H+ + NADH ENSG00000143149
MAR20015    4-aminobutanal + H2O + NAD+     ->  4-aminobutyrate + 2 H+ + NADH ENSG00000143149
MAR06631    NAD+ + retinol  ->  H+ + NADH + retinal ENSG00000025423 or ENSG00000135437 or ENSG00000147576 or ENSG00000170786 or ENSG00000172955 or ENSG00000187758 or ENSG00000196344 or ENSG00000196616 or ENSG00000197894 or ENSG00000198099 or ENSG00000248144
MAR20002    NAD+ + retinol  ->  H+ + NADH + retinal ENSG00000198099 or ENSG00000196616
MAR06646    H2O + NAD+ + retinal    ->  2 H+ + NADH + retinoate ENSG00000128918 or ENSG00000165092
MAR20012    H2O + NAD+ + retinal    ->  2 H+ + NADH + retinoate ENSG00000184254
MAR04545    5-hydroxy-L-tryptophan + H+     ->  CO2 + serotonin ENSG00000132437
MAR20090    5-hydroxy-L-tryptophan + H+     ->  CO2 + serotonin ENSG00000132437
MAR03984    2 H2O2  ->  2 H2O + O2 ENSG00000121691 or ENSG00000166016
MAR20066    2 H2O2  ->  2 H2O + O2 ENSG00000121691
MAR00067    H2O + NAD+ + 2(S)-pristanal     ->  2 H+ + NADH + pristanic acid ENSG00000072210
MAR20052    H2O + NAD+ + 2(S)-pristanal     ->  2 H+ + NADH + pristanic acid ENSG00000072210
> suggest:
> - [ ] remove MAR20020
> - [ ] remove MAR20014
> - [ ] remove MAR20015
> - [ ] remove MAR20002
> - [ ] remove MAR20012
> - [ ] remove MAR20090
> - [ ] remove MAR20066
> - [ ] remove MAR20052

@feiranl very good! it might be better to move this section to a new issue that would likely to help the downstream fixing?

**Comment by haowang-bioinfo** [comment]
fixed in #573, feel free to reopen
